### ISSUE TITLE
Utility for prefetching data on GPUs

### ISSUE THREAD

**Issue by pzelasko** [lhotse-speech/lhotse#243]
My understanding is that with PyTorch's DataLoader, we do the following 

**asynchronously:** 
- I/O reads 
- transforms
- collation

**synchronously:**
- transfer of batches from the worker process to the training process (upon calling `__next__` in the training loop)
- batch transfer from CPU to GPU (upon calling `tensor.to(device)`)

I think we can add a class that wraps the DataLoader (`FastDataLoader`?) and performs both transfer in the background to further speed up the training. There seems a good example here: https://github.com/NVIDIA/apex/blob/master/examples/imagenet/main_amp.py#L265

Another good one (but doesn't address the CPU -> GPU transfer) is in torchaudio: https://github.com/pytorch/audio/blob/master/torchaudio/datasets/utils.py#L276

**Comment by pzelasko** [comment]
... unless I'm missing something - to me it seems like this sort of functionality should be a standard pytorch component?

**Comment by danpovey** [comment]
I suppose so... I don't know whether that transfer is really going to be a
limiting factor, but  if it doesn't make the code significantly
harder to understand I suppose it's a nice thing to have.

**Comment by csukuangfj** [comment]
If I remember correctly, DataLoader in PyTorch does not support GPU when num_workers is greater than one.

**Comment by fanlu** [comment]
@pzelasko Is there a way to store wavefoem in hdf5 format, then we can load data fast and do  on-the-fly augmentation?

**Comment by pzelasko** [comment]
Yeah, it is definitely doable. I don’t know how much faster it would be, but I guess it won’t require to open/close new files all the time. 

As a “quick” workaround you can try increasing the number of dataloader workers.
